### PR TITLE
fix(slider): clicking vertical slider sets value to 0

### LINKED ISSUE #11800: slider(vertical): clicking slider sets value to 0 instead of value clicked when the scroll bar is not at the top
I made a page with vertical `md-slider` inside table. Whenever I scroll down the page, the slider then cannot be clicked or dragged, it will reset it value to minimum like it cannot calculate its value

It seem like it use the page's scroll position to calculate its knob position or something. How can I fix this?

As the picture below. I scroll this page at the top. And then click on the green slider. The value become `61.5` normally

image

But when I just scroll down the page a bit, I try to click the same position on slider. It then become 0 just by clicked

image

### PR DESCRIPTION
**Pull request by Thaina**
Fix #11800 alternative to Fix #11801

## PR Type
What kind of change does this PR introduce?

```
[ x ] Bugfix
[ x ] Enhancement
[ ] Documentation content changes
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: #11800

## What is the new behavior?
Correction for scroll

## Does this PR introduce a breaking change?
```
[ ] Yes
[ ] No
```

### CODE REVIEW

**Comment by Thaina** [comment]
@Splaktar How could we made unittest for something that require visually interaction and environment variable on rendered browser like scrollbar position though?

File: src/components/slider/slider.js [hunk 1]
```diff
@@ -643,15 +643,15 @@ function SliderDirective($$rAF, $window, $mdAria, $mdUtil, $mdConstant, $mdThemi
      * @returns {number}
      */
     function positionToPercent(position) {
-      var offset = vertical ? sliderDimensions.top : sliderDimensions.left;
+      var offset = vertical ? (sliderDimensions.top + window.scrollY) : (sliderDimensions.left + window.scrollX);
```
Comment (Splaktar) [inline src/components/slider/slider.js]: So the idea is that this makes the offset relative to the current viewport (that is scrolled into view) rather than the entire page?

File: src/components/slider/slider.js [hunk 2]
```diff
@@ -643,15 +643,15 @@ function SliderDirective($$rAF, $window, $mdAria, $mdUtil, $mdConstant, $mdThemi
...
       var size = vertical ? sliderDimensions.height : sliderDimensions.width;
       var calc = (position - offset) / size;
 
-      if (!vertical && $mdUtil.isRtl(attr)) {
+      if (vertical || (!vertical && $mdUtil.isRtl(attr))) {
```
Comment (Splaktar) [inline src/components/slider/slider.js]: Can you help me understand why this change is needed? Is this just moving some of the logic from line 654 up to this line?

File: src/components/slider/slider.js [hunk 3]
```diff
@@ -643,15 +643,15 @@ function SliderDirective($$rAF, $window, $mdAria, $mdUtil, $mdConstant, $mdThemi
...
         calc = 1 - calc;
       }
 
-      return Math.max(0, Math.min(1, vertical ? 1 - calc : calc));
+      return Math.min(Math.max(0, calc), 1);
```
Comment (Splaktar) [inline src/components/slider/slider.js]: Would calling the above `clamp(calc)` function make sense here now?

**Comment by Splaktar** [comment]
> @Splaktar How could we made unittest for something that require visually interaction and environment variable on rendered browser like scrollbar position though?

That would require an end to end (e2e) test which we don't currently support in this repo.

However, we can still create unit tests for the `positionToPercent()` function without needing to send any clicks or calculate scroll if we pass in the `scrollX` and `scrollY` as parameters.

File: src/components/slider/slider.js (hunk 1 above)
Comment (Thaina) [reply src/components/slider/slider.js]: @Splaktar No it's opposite. `sliderDimensions.top` and `sliderDimensions.left` is actually relative to current viewport. So we need to `+ scrollY` to make it become relative to the entire page instead. And it would match with `pointer.x`

File: src/components/slider/slider.js (hunk 2 above)
Comment (Thaina) [reply src/components/slider/slider.js]: @Splaktar Because it was the exactly same logic. `x = 1 - x`. It just 2 condition for this logic. `If the slider is vertical` OR `if the slider is horizontal and RTL` it then inverse the value. So I prefer to have same code in the same `if` block

Comment (Thaina) [reply src/components/slider/slider.js]: Maybe we could use exclusive or `^`?

```
if (vertical ^ $mdUtil.isRtl(attr))
```

I felt unfamiliar with XOR in js though

File: src/components/slider/slider.js (hunk 3 above)
Comment (Thaina) [reply src/components/slider/slider.js]: @Splaktar I don't know that we have `clamp` function in this file. The original one also use `Math.max(0, Math.min(1, value))`. I just prefer that if we made a `minmax` in place of `clamp` we should order it in the same manner as `clamp`

**Comment by Thaina** [comment]
> However, we can still create unit tests for the `positionToPercent()` function without needing to send any clicks or calculate scroll if we pass in the `scrollX` and `scrollY` as parameters rather than pulling them from `window` in the function.

@Splaktar could we simulate `window.scroll` like this though?

```js
//// unittest
var tmp = window.scrollY;
window.scrollY = $simValue;
var percent = positionToPercent(position);
window.scrollY = tmp;
```

File: src/components/slider/slider.js (hunk 2 above)
Comment (Splaktar) [reply src/components/slider/slider.js]: No, this is better than XOR. Thank you for the explanation.

**Comment by Splaktar** [comment]
> @Splaktar could we simulate window.scroll like this though?

Probably not in a reliable way. 

If we change the function to `function positionToPercent(position, x, y) {` then we could make it easy to unit test without depending upon the browser's size and ability to scroll the actual browser in a unit test.

**Comment by Thaina** [comment]
Maybe like this?

```js
//// unittest
var percent0 = positionToPercent(position);
var percent1 = positionToPercent(position,{ scrollX : testX,scrollY : testY });
```

**Comment by Splaktar** [comment]
The alternative implementation of this was [merged](https://github.com/angular/material/pull/11801) and released in `1.1.23`.
